src/force-directed-graph/force-directed-graph.component.ts

```diff
@@ -0,0 +1,169 @@
+              [attr.x2]="link.target.x"
+              [attr.y2]="link.target.y"
+            />
+          </svg:g>
+        </svg:g>
+        <svg:g class="nodes">
+          <svg:g *ngFor="let node of nodes"
+            [attr.transform]="'translate(' + node.x + ',' + node.y + ')'"
+            [attr.fill]="colors(groupResultsBy(node))"
+            [attr.stroke]="colors(groupResultsBy(node))"
+            (mousedown)="onDragStart(node, $event)"
+            (click)="click($event, node)"
+
+            swui-tooltip
+            [tooltipPlacement]="'top'"
+            [tooltipType]="'tooltip'"
+            [tooltipTitle]="node.value">
+            <template *ngIf="nodeTemplate"
+              [ngTemplateOutlet]="nodeTemplate"
+              [ngOutletContext]="{ $implicit: node }">
+            </template>
+            <svg:circle *ngIf="!nodeTemplate" r="5" />
+          </svg:g>
+        </svg:g>
+      </svg:g>
+    </chart>
+  `
+})
+export class ForceDirectedGraph extends BaseChart implements OnChanges {
+  colors: Function;
+  dims: ViewDimensions;
+  draggingNode: any;
+  draggingStart: { x: number, y: number };
+  margin = [0, 0, 0, 0];
+  results = [];
+  seriesDomain: any;
+  transform: string;
+
+  @Input() customColors;
+  @Input() force = d3.forceSimulation()
+    .force("charge", d3.forceManyBody())
+    .force("collide", d3.forceCollide(5))
+    .force("x", d3.forceX())
+    .force("y", d3.forceY());
+  @Input() forceLink = d3.forceLink().distance(20).strength(1).id(node => node.value);
+  @Input() groupResultsBy: (node: any) => string = node => node.value;
+  @Input() legend: boolean;
+  @Input() nodes: any[] = [];
+  @Input() links: { source: any, target: any }[] = [];
+  @Input() scheme;
+  @Input() view;
+
+  @Output() clickHandler = new EventEmitter();
+
+  @ContentChild('linkTemplate') linkTemplate: TemplateRef<any>;
+  @ContentChild('nodeTemplate') nodeTemplate: TemplateRef<any>;
+  @ViewChild(Chart, { read: ElementRef }) chart: ElementRef;
+
+  constructor(private element: ElementRef, zone: NgZone) {
+    super(element, zone);
+  }
+
+  ngOnChanges() {
+    this.update();
+  }
+
+  update() {
+    super.update();
+
+    // center graph
+    this.dims = calculateViewDimensions(this.width, this.height, this.margin, false, false, this.legend, 9);
+
+    this.seriesDomain = this.getSeriesDomain();
+    this.setColors();
+
+    this.transform = `translate(${ this.dims.xOffset + this.dims.width / 2 }, ${ this.margin[0] + this.dims.height / 2 })`;
+    if(this.force) {
+      this.force.nodes(this.nodes)
+        .force("link", this.forceLink.links(this.links))
+        .alpha(0.5).restart();
+    }
+  }
+
+  click($event, node) {
+    this.clickHandler.emit(node);
+  }
+
+  getSeriesDomain() {
+    return this.nodes.map(d => this.groupResultsBy(d))
+      .reduce((nodes: any[], node): any[] => nodes.includes(node) ? nodes : nodes.concat([node]), [])
+      .sort();
+  }
+
+  trackBy(index, item) {
```

Comment:
This isn't being used in the template. I would add track by to both nodes and links iterators.